bundle/src/main/java/com/cognifide/cq/cqsm/core/actions/scanner/AnnotatedClassRegistry.java

```diff
@@ -27,96 +35,103 @@
-
-	private final Class<? extends Annotation> annotationClass;
-
-	private final String bundleHeader;
-
-	private final BundleContext bundleContext;
-
-	public AnnotatedClassRegistry(BundleContext bundleContext, String bundleHeader,
-			Class<? extends Annotation> annotationClass) {
-		this.bundleContext = bundleContext;
-		this.bundleHeader = bundleHeader;
-		this.annotationClass = annotationClass;
-
-		tracker = new BundleTracker(bundleContext, Bundle.RESOLVED | Bundle.ACTIVE, this);
-	}
-
-	@Override
-	public Object addingBundle(Bundle bundle, BundleEvent event) {
-		final List<Class<?>> scanned = new ClassScanner(bundle, bundleContext)
-				.findClasses(bundleHeader, annotationClass);
-
-		if (scanned.size() > 0) {
-			if (LOG.isDebugEnabled()) {
-				LOG.debug("Adding classes ({}) from bundle: {}", scanned.size(), bundle.getSymbolicName());
-			}
-
-			classes.put(bundle.getBundleId(), scanned);
-		}
-
-		return null;
-	}
-
-	@Override
-	public void modifiedBundle(Bundle bundle, BundleEvent event, Object object) {
-		final List<Class<?>> scanned = new ClassScanner(bundle, bundleContext)
-				.findClasses(bundleHeader, annotationClass);
-
-		classes.remove(bundle.getBundleId());
-		if (scanned.size() > 0) {
-			if (LOG.isDebugEnabled()) {
-				LOG.debug("Updating classes ({}) from bundle: {}", scanned.size(), bundle.getSymbolicName());
-			}
-
-			classes.put(bundle.getBundleId(), scanned);
-		}
-	}
-
-	@Override
-	public void removedBundle(Bundle bundle, BundleEvent event, Object object) {
-		final List<Class<?>> registered = classes.get(bundle.getBundleId());
-
-		if (LOG.isDebugEnabled()) {
-			LOG.debug("Removing classes ({}) from bundle: {}", registered.size(), bundle.getSymbolicName());
-		}
-
-		classes.remove(bundle.getBundleId());
-	}
-
-	public synchronized List<Class<?>> getClasses() {
-		List<Class<?>> flattened = new ArrayList<>();
-		for (Map.Entry<Long, List<Class<?>>> entry : classes.entrySet()) {
-			flattened.addAll(entry.getValue());
-		}
-
-		return flattened;
-	}
-
-	public void open() {
-		tracker.open();
-	}
-
-	public void close() {
-		tracker.close();
-	}
+public class AnnotatedClassRegistry {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AnnotatedClassRegistry.class);
+
+  private final BundleTracker tracker;
+
+  private final Map<Long, List<Class<?>>> classes = new ConcurrentHashMap<>();
+
+  private final Class<? extends Annotation> annotationClass;
+
+  private final String bundleHeader;
+
+  private final BundleContext bundleContext;
+
+  private final Set<RegistryChangedListener> listeners = new HashSet<>();
+
+  public AnnotatedClassRegistry(final BundleContext bundleContext, final String bundleHeader,
+      final Class<? extends Annotation> annotationClass) {
+    this.bundleContext = bundleContext;
+    this.bundleHeader = bundleHeader;
+    this.annotationClass = annotationClass;
+
+    this.tracker = new BundleTracker(bundleContext, Bundle.ACTIVE, new BundleTrackerCustomizer<Bundle>() {

```

Comment:
I suppose we cannot use lambdas, because it is using java 7? Can we extract this to be in class instead of anonymous class?

Reply:
As you mentioned, that is an anonymous inner class, and such syntax is supported in java from very beginning.